### PR TITLE
[BACKEND] Implementing the Semantic Search Functionality

### LINKED ISSUE #377: [BACKEND] Research Semantic Search
I will research Semantic Search, its adv/disadv over our current search endpoint, and different ways to implement it.

### LINKED ISSUE #442: [BACKEND] Implementing the Semantic Search
I will implement the Semantic Search mechanism.

### PR DESCRIPTION
**Pull request by omerfdeniz**
## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# Description

I implemented the semantic search mechanism to support our filtering/searching mechanism. Previously, products were filtered according to their names, but from now on, semantically similar queries taken from Datamuse API will also be used in search. Each product is indexed in a ProductIndex class where index is preprocessed version of concatenation of product name, descriptions and category info.
* Implemented a mechanism that indexes products when added or updated
* Implemented a Query Parser for Datamuse API
* Implemented preprocessing for text by casefolding, removing punctutation and stopwords and tokenization
* Filtered the products according to similar queries taken from Datamuse API and ranked according to semantic similarity scores
* Then forwarded the results to the filtering mechanism

# Test results

```
(djangoenv) (base) omer-mbp:backend omer$ python manage.py test API.tests.test_search.TestSearchProduct
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
.....
----------------------------------------------------------------------
Ran 5 tests in 0.934s

OK
Destroying test database for alias 'default'...
```
## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have checked my code and corrected any misspellings
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: ecommerce/backend/API/views/search.py [hunk 1]
```diff
@@ -5,22 +5,76 @@
...
+    processed_query = '+'.join(preprocess(query))
+    # get the content of the API
+    r = requests.get(API_URL+processed_query)
+    results = r.content
+    results = json.loads(results)
+    # sort wrt to the highest score which means semantically more similar
+    results = sorted(results, key= lambda x:x['score'], reverse=True)
+    # return words of the top 10 results
+    similar_queries = [d['word'] for d in results[:10]]
+    return similar_queries
```
Comment (bekir96) [inline ecommerce/backend/API/views/search.py]: Excellent work but this function should not represented in search.py

**Review by bekir96 — CHANGES_REQUESTED** [review]
Can you change file direction of relevant function to utils.py?

File: ecommerce/backend/API/views/search.py (hunk 1 above)
Comment (umutoksuz) [reply ecommerce/backend/API/views/search.py]: Aggred. It is better to place it in utils

**Review by umutoksuz — CHANGES_REQUESTED** [review]
Impressive work! If you can move your helper functions to utils folder, it is ready to be merged.

**Review by bekir96 — APPROVED** [review]
All suggestions are satisfied,  you can merge. Excellent work 💯

**Review by furkannane — APPROVED** [review]
Nice work.

File: ecommerce/backend/API/views/search.py (hunk 1 above)
Comment (omerfdeniz) [reply ecommerce/backend/API/views/search.py]: Functions are moved into utils/semantic_search.py

**Review by umutoksuz — APPROVED** [review]
Changes are appropriate. Ready to merge. Very nice work!

**Pull request merged**